### ISSUE TITLE
Regression in 0.3.10 from 0.3.9 parsing optional rfc3339 dates

### ISSUE THREAD

**Issue by lu-zero** [time-rs/time#479]
This test:
``` rust
    #[test]
    fn date() {
        const A: &str = r#"
            {
                "date": "2022-05-01T10:20:42.123Z"
            }
        "#;

        const B: &str = r#"
            {
                "date": "2022-05-01T10:20:42.123Z",
                "maybedate": "2022-05-01T10:20:42.123Z"
            }
        "#;

        #[derive(Debug, Deserialize, Serialize)]
        struct S {
            #[serde(with = "time::serde::rfc3339")]
            date: OffsetDateTime,
            #[serde(with = "time::serde::rfc3339::option", default)]
            maybedate: Option<OffsetDateTime>,
        }

        let a: S = serde_json::from_str(A).unwrap();
        let b: S = serde_json::from_str(B).unwrap();

        println!("{a:?}, {b:?}");
    }
```

passes on `0.3.9` and fails with `0.3.10` with:
```
thread 'thing::test::date' panicked at 'called `Result::unwrap()` on an `Err` value: Error("invalid type: Option value, expected an RFC3339-formatted `OffsetDateTime`", line: 4, column: 29)',
```

**Comment by jhpratt** [comment]
This was introduced in f9398b9, which itself was a fix to #466. It probably would have been caught if serde-rs/serde#2210 was resolved. The ability to test serde implementations without pulling in various serializers/deserializers as dependencies is limited as a result of the serde issue.

I have a possible fix (reverting one line of that commit), but I need to introduce a new test dependency and write a few tests to ensure that nothing regresses. I'd also like to avoid removing the ability to have non-self-describing formats work with `Option<T>` (which is what #466 is about), but I will if it's necessary to avoid this bug.

**Comment by jhpratt** [comment]
I have reverted the relevant line and added a regression test. The partial reversion has just been released as part of 0.3.11.
